Created ControlValues for controlled gates/operations, fix for #4512

NoureldinYosri opened the pull request:
- created control_values.py which contains the ControlValues class.
- FreeVars and ConstrainedVars classes are provided for ease of use.
- while the basic idea of ControlValues integrating it inside the code base was challening
- the old way of using control_values assumed it's a tuple of tuples of ints and was used as thus (comparasion, hashing, slicing, fomatting, conditioning, and loops), the ControlValues class had to provide these functionalities
- the trickiest part to get right was the support for formatting!
- fixes #4512

tanujkhattar commented:
Thanks @NoureldinYosri for opening up the PR. Unit tests need to be added for every new code / modification or else the coverage check won't pass and we can't approve any PRs with failing tests. Please add the unit tests (and fix other failing tests) and then we can do a review for the PR. 

You can look at https://github.com/quantumlib/Cirq/blob/master/docs/dev/development.md#continuous-integration-and-local-testing to learn more about how to run the integration tests locally.

NoureldinYosri commented:
@tanujkhattar sorry for spamming your email with the submissions, I have been running the tests locally but I also had a bash script to do the pull request which was buggy :3, there is also an issue two of the checks are contradictory the Changed files test and Type check want different things in the type hint, I want to use `: ControlValues` but the change file test gives `E   NameError: name 'ControlValues' is not defined` one of them accepts only `Type[ControlValues]`, the other accepts only `Type['ControlValues']` -_-

NoureldinYosri commented:
@tanujkhattar can you check this PR?, all checks pass now :)

dabacon commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,223 @@
+                self.vals = _from_sequence_sequence(val)
+        self.num_variables = len(self.vals[0])
+
+    def And(self, other):  # pylint: disable=invalid-name
```
I assume this is camel case to avoid conflict with `and`?  Suggest using `product`.  That would match the pythonic way to do a cartesian product, `itertools.product`

dabacon commented on cirq-core/cirq/ops/control_values_test.py:
```diff
@@ -0,0 +1,117 @@
+        control_vals = cv.ControlValues(control_values)
+        got = [cv.flatten(product) for product in control_vals]
+        assert want == sorted(got)
+
```
we have a helper for testing equality.  could you and an EqualsTester test?

NoureldinYosri replied:
done

dabacon commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,223 @@
+                    ).format(product[i], i + offset)
+                    if controls is not None:
+                        message = (
+                            'Control values <{product[i]!r}> outside of range'
```
should this be a f-string?

NoureldinYosri replied:
done

dabacon commented on cirq-core/cirq/ops/control_values_test.py:
```diff
@@ -0,0 +1,117 @@
+def test_slicing_not_supported():
+    control_vals = cv.ControlValues([[[0, 1], [1, 0]]])
+    with pytest.raises(ValueError):
+        control_vals[0:1]  # pylint: disable=pointless-statement
```
you can get rid of the pylint warning by doing
`_ = control_vals[0:1]`

NoureldinYosri replied:
done

dabacon commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,223 @@
+        return cur
+
+    def __eq__(self, other):
+        if not isinstance(other, ControlValues):
```
I guess the intention here is to make this equal to other possible ways to build a control value?  If so I think you want to check the types more and if it doesn't equal any of those types, return `NotImplemented` (python's equals contract)

NoureldinYosri replied:
done

dabacon commented on cirq-core/cirq/ops/controlled_gate_test.py:
```diff
@@ -555,3 +555,13 @@ def num_qubits(self) -> int:
             (0.25, cirq.unitary(cirq.CZ)),
         ],
     )
+
```
are we testing the new forms here?  I would add a test for these new cases.

NoureldinYosri replied:
I assume you are talking about mixed gates, no?
can you suggest a test(s)?

daxfohl replied:
I think he means that there are no tests that pass the new class into the ControlledGate constructor and verify the functionality works as desired. There should be tests that do this, preferably with conditions that were not previously expressible.

dabacon reviewed (commented):
Some initial comments.

NoureldinYosri commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,223 @@
+# Copyright 2018 The Cirq Developers
```
done

NoureldinYosri commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,223 @@
+import copy
+import itertools
+
+import cirq  # pylint: disable=unused-import
```
done

NoureldinYosri commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,223 @@
+import cirq  # pylint: disable=unused-import
+
+
+def flatten(sequence):
```
no, flatten works recusively while itertools.chain.from_iterable yields from only the second level, context https://docs.python.org/3/library/itertools.html#itertools.chain.from_iterable

NoureldinYosri commented:
@daxfohl @dabacon this PR has been waiting for sometime, can we get it reviewed :)

daxfohl commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,229 @@
+        self, control_values: Sequence[Union[int, Collection[int], Type['ControlValues']]]
+    ):
+        if len(control_values) == 0:
+            self.vals = cast(Tuple[Tuple[int, ...], ...], (()))
```
Members should be private, with getter properties when necessary.

daxfohl commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,229 @@
+        else:
+            new_copy = ControlValues(
+                [
+                    copy.deepcopy(self.vals),
```
Can this be done without the copy module?

daxfohl replied:
If that's the case, you shouldn't need *any* copy since tuples are immutable. If you remove the copy completely does something break?

daxfohl commented:
Can you add "fixes #4512" to the description so they are linked?

daxfohl commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,229 @@
+            self.nxt = None
+            self.itr = None
+            return
+        self.itr = None
```
Does this need to be a member? Frequently classes that iterate via member iterators have problems when iterating multiple times, especially with multiple iterations going on at once. It's usually better to create the iterators independently in the methods that do the iteration.

daxfohl reviewed (commented):
One more thing

NoureldinYosri commented:
I addressed most of the comments, I have a couple of questions which I put in the comments.

daxfohl commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,229 @@
+    return tuple(tuple(product) for product in vals)
+
+
+class ControlValues:
```
The root `__init__.py` file, so users can do `cirq.ControlValues`. (Once you do this you'll find it triggers some additional tests that check for serializability etc. These tasks need to be completed too, or else the control won't be able to run on real machines).

daxfohl commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,272 @@
+            raise TypeError(f'{val} is of Unsupported type {type(val)}')
+        self._num_variables = len(self._vals[0])
+
+    def product(self, other):
```
All public methods need type annotations

NoureldinYosri replied:
added

daxfohl commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,272 @@
+        cur = self
+        while cur._nxt is not None:
+            cur = cur._nxt
+        cur._nxt = other
```
This needs to be implemented in a functional fashion. Cirq gates are supposed to be immutable (this has been overlooked in places, but there's an effort to clean those up). This modifies the `ControlValues` object in-place, which given it's a member of a gate, provides a backdoor to mutating the gate.

NoureldinYosri replied:
done

daxfohl commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,272 @@
+        """Removes the last control values combination."""
+        if self._nxt is None:
+            return None
+        self._nxt = self._nxt.pop()
```
Again, needs to be implemented functionally, not in-place modification.

daxfohl commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,272 @@
+            return itertools.product(*(), nxt())
+
+    def copy(self):
+        """Returns a deep copy of the object."""
```
Once everything has been implemented functionally, there should be no need for this function.

NoureldinYosri replied:
done

daxfohl commented:
@NoureldinYosri consider using `@dataclass(frozen=True)` on this to help ensure immutability. We use that elsewhere in cirq, you can just search for examples. It provides nice default implementation of equality, hash, etc as well. Just an option--if it gets in the way then don't worry about it.

NoureldinYosri commented:
@daxfohl @dabacon @tanujkhattar ptal

daxfohl commented:
The changes resolve my concerns about mutability. Still need to add these classes to the root namespace in `cirq-core/cirq/__init__.py`. Also the classes' top level docstrings need some more detailed explanation and examples suitable for end users. Beyond that, I think I'll pass this along to the maintainers at this point.

MichaelBroughton commented:
There are some concerns here re: qsim compatability @95-martin-orion , this has implications for TFQ as well and our goals for decompose always returning X,Y,Z,C*Z. Have you looked into this Orion ?

95-martin-orion reviewed (changes requested):
Quick review pass + follow up to Mike's question

NoureldinYosri commented on cirq-core/cirq/ops/controlled_operation.py:
```diff
@@ -134,7 +141,8 @@ def _extend_matrix(self, sub_matrix: np.ndarray) -> np.ndarray:
         tensor = qis.eye_tensor(qid_shape, dtype=sub_matrix.dtype)
         sub_tensor = sub_matrix.reshape(qid_shape[len(self.controls) :] * 2)
-        for control_vals in itertools.product(*self.control_values):
+        for control_vals in self.control_values:
+            control_vals = cv.ControlValues.flatten(control_vals)
```
the `|00><00| + |11><11|` combination can be added in the same way that the xor combination `|01><01| + |10><10|` which was not supported in the old way and can be seen in the tests

daxfohl replied:
Did we ever have a design review on this one? I provided the initial code reviews assuming a design was approved, but I still kind of stand by what I originally said on the linked issue, that this could be done more simply by just adding a bool flag to the ControlledGate constructor. That flag would control whether to interpret the 2D input array vertically as allowed states by qubit (which is how it currently works, but is limited), or horizontally as a full set of allowed states (which allows full control). 

That this question came up during code review makes me think end users are going to have trouble understanding how to use these classes.

NoureldinYosri replied:
I introduced the idea in https://github.com/quantumlib/Cirq/issues/4512#issuecomment-952284810 and talked about it ina cirq meeting https://github.com/quantumlib/Cirq/issues/4512#issuecomment-953208347, wrote a prototype in https://github.com/quantumlib/Cirq/issues/4512#issuecomment-954142591 and got a review in https://github.com/quantumlib/Cirq/issues/4512#issuecomment-974322956

--------------------------------------------------------------
the boolean idea has some issues
1. it doesn't address your previous comment  https://github.com/quantumlib/Cirq/issues/4512#issuecomment-1014129817, it fact it will make it harder to add those new ways lalter.
2. there is no way of getting around computing products, what a boolean will do is to force the users to compute them before passing them to the constructor, a work around would be passing an array of booleans but that has its own issues.
3. any work around will make the constructors for controlledGate and controlledOperation execture complex logic which is not a good idea in general, introduce tech debt and make it harder change or support new types later,

-----------------------------------------------------------------
lets discuss this in the next cirq meeting.

daxfohl replied:
Sounds good. I have limited availability for cynqs these days, but I think a reprise with the rest of the team on Wed would be great just to get the ball rolling on this PR, since it seems to keep going stagnant. I'm looking forward to the functionality.

NoureldinYosri commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,447 @@
+class ControlValues(abc.ABC):
+    """ControlValues is an abstract data class that represents different control values."""
+
+    _combinations: Any
+    _nxt: Optional['ControlValues'] = None
```
this assumes that lists of combinations are the only way that wil be used to define control_values, there was a previous comment on this PR about possibly supporting other ways like tensors or sympy expressions in the future.

the linked-list structure allows defining using different ways at the same time e.g. CV1(a table) -> CV2(a sympy expression) -> ... etc

each of those object will inheret from the ControlValues base class so that wil work perfectly.

daxfohl replied:
The comment referred to above is on the linked issue. https://github.com/quantumlib/Cirq/issues/4512#issuecomment-1014129817

daxfohl replied:
Another thought, have you considered making a separate `ControlValuesProduct` class that merges two (or maybe N) ControlValues objects into a product? That approach would allow for a smaller base class and adhere more to the single responsibility pattern.

NoureldinYosri reviewed (commented):
I'll remove the debug line and add the classes to the root namespace ASAP.

the main point of the link-list structure is all for different ways of creating  control_values (e.g. the old way, the new way, using tensors, using generators, using sympy expressions ..etc) to be used together.

95-martin-orion commented:
Bump: the checks appear to have gotten stuck 😖 . @NoureldinYosri, you may need to push a change (maybe resolving the conflicts marked) to get it moving again.

daxfohl commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -0,0 +1,453 @@
+
+
+@ControlValues.register
+class SimpleControlValues(ControlValues):
```
Let's bikeshed this during synq. I don't think users will find the name intuitive. Perhaps `PerQubitControlValues`? Also the docstring could use some examples to make it clear.

daxfohl reviewed (commented):
Added a couple of thoughts that we could go over during cynq.

daxfohl commented on cirq-core/cirq/ops/controlled_gate.py:
```diff
@@ -241,26 +247,26 @@ def get_prefix(control_vals):
         return (
             f'cirq.ControlledGate(sub_gate={self.sub_gate!r}, '
-            f'control_values={self.control_values!r},'
+            f'control_values={self.control_values.arrangements()!r},'
```
Does this still round-trip? Given there's no type information about which subclass of `ControlValues` it uses, I'm not seeing how it could.

daxfohl reviewed (commented):
One more question

NoureldinYosri commented:
@daxfohl @95-martin-orion I have a simple way of implementing this which I think addresses all the comments/concerns, I will present it during the cirq meeting.

daxfohl commented:
Great! One other thing to look out for is the circuit diagram. The current
logic labels each qubit with the control values. That probably won't work
for the new scenarios. Will need a different approach for them.

On Mon, Mar 21, 2022, 6:46 AM Noureldin ***@***.***> wrote:

> @daxfohl <https://github.com/daxfohl> @95-martin-orion
> <https://github.com/95-martin-orion> I have a simple way of implementing
> this which I think addresses all the comments/concerns, I will present it
> during the cirq meeting.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/quantumlib/Cirq/pull/4714#issuecomment-1073917557>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG464PGUDURYJ2V5NQRFTLVBB4VDANCNFSM5JCVGBLA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

95-martin-orion commented:
Discussed on Cirq sync: @NoureldinYosri will write a document capturing the overarching design of this change, we'll discuss and reach a consensus there, then any changes can be reflected here.

NoureldinYosri commented:
As discussed in the cirq-sync meeting I wrote a [design](https://docs.google.com/document/d/1TtHd4Cfyo9Zwrxtols5N_Ffv_dbXH2XgBB6D3tNIUYo/edit?usp=sharing&resourcekey=0-Zwlj47ZlUM626InJWsvWiA) descriping the current approach.

@tanujkhattar @95-martin-orion @daxfohl @dabacon

tanujkhattar commented:
This can now be closed after https://github.com/quantumlib/Cirq/pull/5788
